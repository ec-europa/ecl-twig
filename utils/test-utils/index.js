const Twig = require('twig');
const { format } = require('prettier');

const renderTwigFile = (file, options, cb) =>
  Twig.renderFile(file, options, (err, html) =>
    cb(err, format(html, { parser: 'html' }))
  );

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) =>
    Twig.renderFile(file, options, (err, html) => {
      if (err) return reject(err);

      const div = document.createElement('div');
      div.innerHTML = html.trim();

      return resolve(div.firstChild);
    })
  );

module.exports = {
  renderTwigFile,
  renderTwigFileAsNode,
};
