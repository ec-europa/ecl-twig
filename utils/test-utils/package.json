{
  "name": "@ecl-twig/test-utils",
  "version": "2.29.0",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
