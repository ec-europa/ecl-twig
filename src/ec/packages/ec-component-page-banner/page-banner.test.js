import path from 'path';
import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';

import bannerDataDefault from '@ecl/ec-specs-page-banner/demo/data--default';
import bannerDataImage from '@ecl/ec-specs-page-banner/demo/data--image';
import bannerDataImageShade from '@ecl/ec-specs-page-banner/demo/data--image-shade';
import bannerDataPrimary from '@ecl/ec-specs-page-banner/demo/data--primary';
import bannerDataAlignLeft from '@ecl/ec-specs-page-banner/demo/data--align-left';

function formatBanner(b) {
  const iconType = b.button.icon.shape.split('--');
  const banner = {
    type: b.variant,
    title: b.title,
    description: b.description,
    button: {
      variant: b.button.variant,
      label: b.button.label,
      icon: {
        type: iconType[0],
        name: iconType[1],
        transform: b.button.icon.transform,
        size: b.button.icon.size,
        path: 'static/icons.svg',
      },
    },
    centered: b.isCentered,
  };

  if ('image' in b) {
    banner.image = 'https://v2--europa-component-library.netlify.com'.concat(
      b.image
    );
  }

  return banner;
}

describe('EC - Page Banner', () => {
  const template = path.resolve(__dirname, './page-banner.html.twig');
  const render = params => renderTwigFileAsNode(template, params);

  describe('default', () => {
    const data = formatBanner(bannerDataDefault);

    test(`- renders correctly`, () => {
      expect.assertions(1);

      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('image', () => {
    const data = formatBanner(bannerDataImage);

    test(`- renders correctly`, () => {
      expect.assertions(1);

      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('image-shade', () => {
    const data = formatBanner(bannerDataImageShade);

    test(`- renders correctly`, () => {
      expect.assertions(1);

      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('primary', () => {
    const data = formatBanner(bannerDataPrimary);

    test(`- renders correctly`, () => {
      expect.assertions(1);

      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('align-left', () => {
    const data = formatBanner(bannerDataAlignLeft);

    test(`- renders correctly`, () => {
      expect.assertions(1);

      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });
});
