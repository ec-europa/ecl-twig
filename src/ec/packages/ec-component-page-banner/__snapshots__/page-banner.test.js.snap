// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Page Banner align-left - renders correctly 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner align-left renders correctly with extra attributes 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner align-left renders correctly with extra class names 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default custom-class custom-class--test"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner default - renders correctly 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default ecl-page-banner--centered"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner default renders correctly with extra attributes 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default ecl-page-banner--centered"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner default renders correctly with extra class names 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--default ecl-page-banner--centered custom-class custom-class--test"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image - renders correctly 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image ecl-page-banner--centered"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image renders correctly with extra attributes 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image ecl-page-banner--centered"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image renders correctly with extra class names 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image ecl-page-banner--centered custom-class custom-class--test"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image-shade - renders correctly 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image-shade ecl-page-banner--centered"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image-shade renders correctly with extra attributes 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image-shade ecl-page-banner--centered"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner image-shade renders correctly with extra class names 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--image-shade ecl-page-banner--centered custom-class custom-class--test"
  >
    <div
      class="ecl-page-banner__image"
      style="background-image:url(https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg)"
    />
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner primary - renders correctly 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--primary ecl-page-banner--centered"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner primary renders correctly with extra attributes 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--primary ecl-page-banner--centered"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;

exports[`EC - Page Banner primary renders correctly with extra class names 1`] = `
<test>
  <section
    class="ecl-page-banner ecl-page-banner--primary ecl-page-banner--centered custom-class custom-class--test"
  >
    <div
      class="ecl-container ecl-page-banner__container"
    >
      <div
        class="ecl-page-banner__content"
      >
        <h1
          class="ecl-page-banner__title"
        >
          EU Budget for the future
        </h1>
        <p
          class="ecl-page-banner__baseline"
        >
          Innovation, economy, environment and geopolitics
        </p>
        <a
          class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after ecl-page-banner__link-cta"
          href="/example"
        >
          <span
            class="ecl-link__label"
          >
            Subscribe
          </span>
           
          <svg
            aria-hidden="true"
            class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-link__icon"
            focusable="false"
          >
            <use
              xlink:href="/icons.svg#ui--corner-arrow"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</test>
`;
